### ISSUE TITLE
bayesflow w/ higher order functions loops endlessly

### ISSUE THREAD

**Issue by j-wilson** [tensorflow/tensorflow#10968]
**Environment**: TensorFlow v1.2.0-rc2-21-g12f033d 1.2.0 running on Mac OS X (v10.12.5)

**Issue**:
Use of `tensorflow.contrib.bayesflow.stochastic_tensor.StochasticTensor` in conjunction with higher order functions, such as `tf.map_fn` and `tf.while_loop`, results in a seemingly endless loop during construction of the associated stochastic graph's `surrogate_loss`. Within the source code, the aforementioned loop occurs [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/bayesflow/python/ops/stochastic_graph_impl.py#L85). A minimal example is provided below.

**Example**:
```
# Dependencies
import tensorflow as tf
from tensorflow.contrib.distributions import Bernoulli
from tensorflow.contrib.bayesflow import\
	stochastic_tensor as st, stochastic_graph as sg

# Flags and variable declaration
_int, _float = 'int32', 'float32'
num_logits = 0 #doesn't seem to matter
logits = tf.get_variable('logits', shape=[num_logits, 1], dtype=_float)

# Pass `logits` to a distribution and then wrap with a StochasticTensor
dist = Bernoulli(logits, dtype=_float)
with st.value_type(st.SampleValue()):
	samples = st.StochasticTensor(dist)
```
Building off of the above, the following seem to loop endlessly. Using `tf.map_fn`,
```
losses = tf.map_fn(lambda x: x, samples, _float)
loss = sg.surrogate_loss([tf.reduce_mean(losses, axis=0)])# <- loops endlessly
```
Using `tf.while_loop` in conjunction with `tf.TensorArray`,
```
array = tf.TensorArray(_float, num_logits).unstack(samples)
def cond(k, *args):
	return tf.less(k, num_logits)
def body(k, losses):
	return k+1, tf.concat([losses, array.read(k)], 0)
loop_v = [0, tf.zeros([0], _float)]
invars = [tf.TensorShape([]), tf.TensorShape([None])]
losses = tf.while_loop(cond, body, loop_v, shape_invariants=invars)[1]
loss = sg.surrogate_loss([tf.reduce_mean(losses, keep_dims=True)])# <- loops endlessly
```

**Comment by jvdillon** [comment]
Hello and thanks for raising this issue.

Unfortunately control flow primitives are difficult to support with StochasticTensors.  Currently, the best solution is to sample and calculate log_prob at every timestep within the while_loop and emit both; then calculate the loss yourself.
